csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
+        // references. we need both ref candidates to have the same mapping on
+        // allocation domain
+        for (auto i : c10::irange(ref->nDims())) {
+          if (!val_sets.permissiveAreMapped(
+                  ref->getMaybeAllocationDomain()[i],
+                  iter.first->getMaybeAllocationDomain()[i])) {
+            // reset ref to nullptr, while keeping the iterdomain count high
+            // water mark. No propagatoin will occur unless we found another ref
+            // candidate with a higher iterdomain count.
+            ref = nullptr;
+            break;
+          }
+        }
+        continue;
+      }
+    }
 
-  // Note: we only consider simple permutation of allocation domain to rfactor
-  // domain.
-  for (auto tv : ir_utils::filterByType<TensorView>(fusion->inputs())) {
-    std::optional<AllocationOrder> permutation = ir_utils::computePermutation(
-        TensorDomain::noReductions(tv->getMaybeRFactorDomain()),
-        TensorDomain::noReductions(tv->getMaybeAllocationDomain()));
-    if (permutation.has_value()) {
-      alloc_order_map[tv] = permutation.value();
+    // propagate allocation domain if we still have a candidate.
+    if (ref) {
+      AllocationOrderMapping(id_model, ref, dst);
     }
   }
-
-  // Initialize AllocationOrderInferencer with allocation order of input tensor
-  // views
-  AllocationOrderInferencer infer(alloc_order_map);
-  infer.traverse(fusion);
-
-  // return the propagated map
-  return alloc_order_map;
 }
 
 void AllocationDomainPass::runPass(Fusion* fusion) {
-  std::unordered_map<const TensorView*, AllocationOrder> stride_mapping =
-      inferenceAllocationOrder(fusion);
-
-  for (Val* out_val : fusion->outputs()) {
-    auto* out_tv = dynamic_cast<TensorView*>(out_val);
-    // skip:
-    //   1. non-tensor output;
-    //   2. tensor output with allocation specified, assuming everything is
-    //   semantical
-    //   3. tensor output that's aliasing (Does aliased src matter?)
-    if (out_tv == nullptr || out_tv->hasAllocation() ||
-        fusion->getOutputAlias(out_val).type != AllocationType::New) {
-      continue;
-    }
-
-    auto mapped_entry = stride_mapping.find(out_tv);
-    if (mapped_entry == stride_mapping.end() || mapped_entry->second.empty()) {
-      continue;
-    }
-
-    out_tv->setAllocationDomain(
-        constructAllocationDomain(out_tv, mapped_entry->second), true);
-  }
+  // mark input TensorViews as propagation sources
+  auto input_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+  std::vector<TensorView*> srcs(input_tvs.begin(), input_tvs.end());
+  // mark output TensorViews as propagation destinations
+  auto output_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+  std::vector<TensorView*> dsts(output_tvs.begin(), output_tvs.end());
+  // propagate allocation domain from sources to destinations
+  inferenceAllocationOrder(fusion, srcs, dsts);
```

Comment:
Given `srcs` and `dsts` can be determined by reading `fusion`, why do we pass them as arguments? Because you need to customize that in tests?

Reply:
not just in tests, but also as a utility thing to use. i.e. we would also want to include intermediate tensor. Immediate usage is for inputs to matmul/linear. See #2198 and https://github.com/NVIDIA/Fuser/pull/2168#issuecomment-2099086932